src/runtime/matlab.cpp

```diff
@@ -0,0 +1,377 @@
+
+// Given a halide type code and bit width, find the equivalent matlab class ID.
+WEAK mxClassID get_class_id(int32_t type_code, int32_t type_bits) {
+    switch (type_code) {
+    case halide_type_int:
+        switch (type_bits) {
+        case 1: return mxLOGICAL_CLASS;
+        case 8: return mxINT8_CLASS;
+        case 16: return mxINT16_CLASS;
+        case 32: return mxINT32_CLASS;
+        case 64: return mxINT64_CLASS;
+        }
+        return mxUNKNOWN_CLASS;
+    case halide_type_uint:
+        switch (type_bits) {
+        case 1: return mxLOGICAL_CLASS;
+        case 8: return mxUINT8_CLASS;
+        case 16: return mxUINT16_CLASS;
+        case 32: return mxUINT32_CLASS;
+        case 64: return mxUINT64_CLASS;
+        }
+        return mxUNKNOWN_CLASS;
+    case halide_type_float:
+        switch (type_bits) {
+        case 32: return mxSINGLE_CLASS;
+        case 64: return mxDOUBLE_CLASS;
+        }
+        return mxUNKNOWN_CLASS;
+    }
+    return mxUNKNOWN_CLASS;
+}
+
+// Convert a matlab class ID to a string.
+WEAK const char *get_class_name(mxClassID id) {
+    switch (id) {
+    case mxCELL_CLASS: return "cell";
+    case mxSTRUCT_CLASS: return "struct";
+    case mxLOGICAL_CLASS: return "logical";
+    case mxCHAR_CLASS: return "char";
+    case mxVOID_CLASS: return "void";
+    case mxDOUBLE_CLASS: return "double";
+    case mxSINGLE_CLASS: return "single";
+    case mxINT8_CLASS: return "int8";
+    case mxUINT8_CLASS: return "uint8";
+    case mxINT16_CLASS: return "int16";
+    case mxUINT16_CLASS: return "uint16";
+    case mxINT32_CLASS: return "int32";
+    case mxUINT32_CLASS: return "uint32";
+    case mxINT64_CLASS: return "int64";
+    case mxUINT64_CLASS: return "uint64";
+    case mxFUNCTION_CLASS: return "function";
+    case mxOPAQUE_CLASS: return "opaque";
+    case mxOBJECT_CLASS: return "object";
+    default: return "unknown";
+    }
+}
+
+// Declare function pointers for the mex APIs.
+#define MEX_FN(ret, func, args) ret (*func)args;
+#ifndef BITS_32
+# define MEX_FN_730(ret, func, func_730, args) MEX_FN(ret, func, args)
+#else
+# define MEX_FN_700(ret, func, func_700, args) MEX_FN(ret, func, args)
+#endif
+#include "mex_functions.h"
+
+// Get the real data pointer from an mxArray.
+template <typename T>
+INLINE T* get_data(mxArray *a) { return (T *)mxGetData(a); }
+template <typename T>
+INLINE const T* get_data(const mxArray *a) { return (const T *)mxGetData(a); }
+
+// Search for a symbol in the calling process (i.e. matlab).
+template <typename T>
+INLINE T get_symbol(void *user_context, const char *name) {
+    T s = (T)find_symbol(name);
+    if (s == NULL) {
+        error(user_context) << "Matlab API not found: " << name << "\n";
+        return NULL;
+    }
+    return s;
+}
+
+}  // namespace mex
+}  // namespace Runtime
+}  // namespace Halide
+
+using namespace Halide::Runtime::mex;
+
+extern "C" {
+
+WEAK void halide_matlab_error(void *, const char *msg) {
```

Comment:
They'd have to be weak, and there's no way to guarantee that the final binary would contain the original version or these. Also, we don't want to call mexPrintf until halide_matlab_init has run. (will add a comment on this).